akka-http/src/main/scala/akka/http/javadsl/server/directives/HeaderDirectives.scala

```diff
@@ -61,10 +66,31 @@ abstract class HeaderDirectives extends FutureDirectives {
    * If no header with a matching type is found the request is rejected with a [[akka.http.javadsl.server.MissingHeaderRejection]].
    */
   def headerValueByType[T <: HttpHeader](t: Class[T], inner: jf.Function[T, Route]) = RouteAdapter {
-    // TODO custom headers don't work yet
-    // TODO needs instance of check if it's a modeled header and then magically locate companion
-    D.headerValueByType(HeaderMagnet.fromClassNormalJavaHeader(t)) { value =>
-      inner.apply(value).delegate
+
+    if (classOf[ModeledCustomHeader[_]].isAssignableFrom(t)) {
+      // figure out the modeled header companion and use that to parse the header
+      val refl = new ReflectiveDynamicAccess(getClass.getClassLoader)
+      val magnet: HeaderMagnet[_] =
+        refl.getObjectFor[ModeledCustomHeaderCompanion[_]](t.getName) match {
+          case Success(companion) =>
+            new HeaderMagnet[T] {
+              override def classTag = ClassTag(t)
+              override def runtimeClass = t
+              override def extractPF = {
+                case h if h.is(companion.lowercaseName) => companion.apply(h.toString).asInstanceOf[T]
+              }
+            }
```

Comment:
Couldn't figure out how to lye enough convincingly to scalac about being sure the type parameter `T <: ModeledCustomHeader[T]` so I couldn't use `HeaderMagnet.fromClassForModeledCustomHeader` here